### PR TITLE
Build-system overhaul

### PR DESCRIPTION
**Pull request by bgrube**
This is based work by @suhlatwork . This PR includes and supercedes #227 and #239 and also includes ab731a35078583d766635b51189f2b5f9fc2c78f from #222.

**Main changes:**
* Increased minimum required versions of a number of dependencies:
    * CMake 3.11.0
    * Python 2.7
    * Boost 1.67.0
        * Use new CMake feature to specify optional Boost components
        * Switch to new Boost.Python library naming scheme
    * yaml-cpp 0.6.0
    * ROOT 6.08.00
    * NLopt 2.4.1
    * ccache 3.1.9
* Updated README accordingly
* CMake build system
    * Scripts now detect and auto-enable ccache to speed-up compilation runs
    * Code cleanup in FindXXX.cmake scripts
        * Unified handling of exported variables
        * Where useful scripts now use find_package_handle_standard_args() function
        * Added version check to FindNLopt.cmake
        * FindPython.cmake now limits Python version to 2.x
    * Improved reporting
    * Added common macros for version string parsing
    * Compilation of Fortran code is now disabled if no Fortran compiler is found
    * Improved and fixed environment reporting
    * Added option to override C++ standard used by compiler
* Build script for Boost now has a switch to redirect output either to stdout or to log file
* Improved and extended Travis build
    * Added clang-format run to build matrix
    * Enabled ccache
    * Enable parallel compilation jobs to reduce run time
    * Centralized version definition of dependencies in Travis config file; unified install scripts
    * Improved modularization of Travis shell scripts
    * Improved cleaning of cached directory where dependencies are installed
* Fix some compilation warnings that appeared with gcc 8.2
* Successfully tested with CERN lxplus LCG_95/x86_64-slc6-gcc8-opt environment
   See <https://ep-dep-sft.web.cern.ch/afs-phaseout> and <http://lcginfo.cern.ch/>

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+AlignConsecutiveAssignments: true
+AlignConsecutiveDeclarations: true
+AlignEscapedNewlinesLeft: true
+AlignOperands: true
+AlignTrailingComments: true
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: false
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: Inline
+AllowShortIfStatementsOnASingleLine: false
```
[the remark this answers is not in the document]
Comment (bgrube) [reply .clang-format]: Changed to `AllowShortIfStatementsOnASingleLine: true`

File: CMakeLists.txt [hunk 2]
```diff
@@ -152,12 +150,65 @@ endforeach()
...
+				message(STATUS "ccache seems to be already in use for ${_LANG}.")
+			else()
+				message(STATUS "Enabling ccache for ${_LANG}")
+				set(CMAKE_${_LANG}_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE} CACHE STRING "")
+			endif()
+			unset(_LANG)
+			unset(_COMPILER_PATH)
+		endforeach()
+	endif()
+endif()
```
[the remark this answers is not in the document]
Comment (bgrube) [reply CMakeLists.txt]: Implemented version check. Require minimal ccache version of 3.1.9 as is provided by Travis Trusty environment.

File: CMakeLists.txt [hunk 3]
```diff
@@ -229,6 +268,7 @@ set_package_properties(Libconfig
 	PROPERTIES
 	DESCRIPTION "Simple C++ library for processing structured configuration files"
 	URL "http://www.hyperrealm.com/libconfig/"
+	PURPOSE "Used for defining partial-wave decay amplitudes"
```
[the remark this answers is not in the document]
Comment (bgrube) [reply CMakeLists.txt]: Done

File: CMakeLists.txt [hunk 4]
```diff
@@ -237,50 +277,29 @@ set_package_properties(Libconfig
 # environment variable $YAML_CPP is expected to point to yaml-cpp location
 message(STATUS "")
 message(STATUS ">>> Setting up yaml-cpp library.")
-find_package(YamlCpp 0.5 REQUIRED)
+find_package(YamlCpp 0.6.0 REQUIRED)
 set_package_properties(YamlCpp
 	PROPERTIES
 	DESCRIPTION "Simple C++ library for processing structured configuration files"
 	URL "https://github.com/jbeder/yaml-cpp"
+	PURPOSE "Used for defining analysis environments"
```
[the remark this answers is not in the document]
Comment (bgrube) [reply CMakeLists.txt]: Done.

File: cmakeModules/CommonMacros.cmake [hunk 5]
```diff
@@ -185,3 +185,35 @@ function(var_name_from_dir_name SRC_DIR VAR_NAME)
...
+# helper function that parses version string from given pkg-config's .rc file
+# this done using the CMake facilities
+# an alternative approach would be to use the output of the pkg-config executable
+function(parse_version_from_pkg_config_file RC_FILE VERSION)
+	set(${VERSION} "NOTFOUND" PARENT_SCOPE)
+	if(NOT EXISTS "${RC_FILE}")
+		message(WARNING "pkg-config file '${RC_FILE}' does not exist. Cannot parse version.")
+	else()
+		file(STRINGS ${RC_FILE} _VERSION_LINE REGEX "Version:")
+		list(LENGTH _VERSION_LINE _VERSION_LINES)
```
[the remark this answers is not in the document]
Comment (bgrube) [reply cmakeModules/CommonMacros.cmake]: Yes, much better.

File: cmakeModules/FindBAT.cmake [hunk 6]
```diff
@@ -49,7 +49,12 @@ if(NOT BAT_ROOT_DIR)
 
 endif()
 
-if(BAT_ROOT_DIR)
+
+if(NOT BAT_ROOT_DIR)
+	set(BAT_FOUND FALSE)
+	set(BAT_ERROR_REASON "${BAT_ERROR_REASON} cannot find BAT directory. "
+		"Either environment variable BAT_ROOT_DIR is not set correctly or bat-config is not in path.")
```
[the remark this answers is not in the document]
Comment (bgrube) [reply cmakeModules/FindBAT.cmake]: It is set to empty string on line 25: https://github.com/ROOTPWA-Maintainers/ROOTPWA/blob/0b41509cd060e3d9d4d419fec975deac6b4328a6/cmakeModules/FindBAT.cmake#L25

File: cmakeModules/FindLibconfig.cmake [hunk 7]
```diff
@@ -41,140 +41,129 @@
...
+	get_filename_component(Libconfig_LIBRARY_DIR ${Libconfig_LIBS} DIRECTORY)
+else()
+	set(Libconfig_FOUND FALSE)
+	if(Libconfig_DIR)
+		set(Libconfig_ERROR_REASON "${Libconfig_ERROR_REASON} Did not find Libconfig library '${_Libconfig_LIBRARY_NAMES}' "
+			"in directory '${Libconfig_DIR}/lib'.")
 	else()
-		get_filename_component(Libconfig_DIR ${Libconfig_LIBS} PATH)
+		set(Libconfig_ERROR_REASON "${Libconfig_ERROR_REASON} Did not find Libconfig library '${_Libconfig_LIBRARY_NAMES}' "
+			"in any standard library directory.")
```
[the remark this answers is not in the document]
Comment (bgrube) [reply cmakeModules/FindLibconfig.cmake]: The idea is to use this pattern whenever there is more than one failure mode.

File: cmakeModules/FindNLopt.cmake [hunk 8]
```diff
@@ -37,98 +38,134 @@
...
 else()
+	# search system-wide
+	find_file(_NLopt_PC_FILE
+		NAMES ${_NLopt_PKG_CONFIG_FILE_NAME}
+		PATHS ${NLopt_LIBRARY_DIR}/pkgconfig
+		      ${NLopt_LIBRARY_DIR}
+		#NO_DEFAULT_PATH
+		)
+endif()
+unset(_NLopt_PKG_CONFIG_FILE_NAME)
```
[the remark this answers is not in the document]
Comment (bgrube) [reply cmakeModules/FindNLopt.cmake]: Done.

File: cmakeModules/FindNLopt.cmake [hunk 9]
```diff
@@ -37,98 +38,134 @@
...
-	)
-
-
 # report result
 if(NLopt_FOUND)
-	message(STATUS "Found NLopt in '${NLopt_DIR}'.")
+	message(STATUS "Found NLopt version ${NLopt_VERSION} in '${NLopt_DIR}'.")
 	message(STATUS "Using NLopt include directory '${NLopt_INCLUDE_DIR}'.")
 	message(STATUS "Using NLopt library '${NLopt_LIBS}'.")
 else()
```
[the remark this answers is not in the document]
Comment (bgrube) [reply cmakeModules/FindNLopt.cmake]: Further unified FindXXX scripts:
- now all except FindROOT use CMake's find_package_handle_standard_args() macro to handle find_package arguments and version checks
- all scripts unset all exported variables if package was not found

File: travisCI/install-bat.sh [hunk 10]
```diff
@@ -1,34 +1,35 @@
...
-	make
+	make --jobs=${NMB_JOBS}
 	make install
 
-	cd ${TRAVIS_BUILD_DIR}/deps/
+	cd "${TRAVIS_BUILD_DIR}"/deps/
 fi
-ln -sfn bat-${BAT_REVISION} bat
+
+ln -sfn bat-${BAT_VERSION} bat
```
[the remark this answers is not in the document]
Comment (bgrube) [reply travisCI/install-bat.sh]: Done.

File: cmakeModules/FindLibconfig.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefan-wallner) [reply cmakeModules/FindLibconfig.cmake]: I fully agree. Please keep it like that.

File: test/decayAmplitude/testAmplitudeTree.cc [hunk 11]
```diff
@@ -95,7 +88,7 @@ main()
 		inFile->GetObject("test", tree);
 		amplitudeTreeLeaf* ampLeaf = 0;
 		tree->SetBranchAddress("amp", &ampLeaf);
-		for (unsigned int i = 0; i < tree->GetEntriesFast(); ++i) {
+		for (long i = 0; i < tree->GetEntriesFast(); ++i) {
```
Comment (stefan-wallner) [inline test/decayAmplitude/testAmplitudeTree.cc]: `GetEntriesFast` returns `Long64_t`, which is `long long` not `long`. Those two are different on Windows systems (see [here](https://de.cppreference.com/w/cpp/language/types)).
I would suggest to use the ROOT type `Long64_t`.

File: CMakeLists.txt [hunk 12]
```diff
@@ -250,17 +250,18 @@ message(STATUS "Using python bindings output path '${PYTHON_LIBRARY_OUTPUT_PATH}
 # environment variable $BOOST_ROOT is expected to point to non-standard locations
 message(STATUS "")
 message(STATUS ">>> Setting up Boost library.")
+set(Boost_USE_STATIC_LIBS      OFF)
+set(Boost_USE_MULTITHREADED    ON)
+set(Boost_USE_STATIC_RUNTIME   OFF)
+set(Boost_DETAILED_FAILURE_MSG ON)
```
Comment (stefan-wallner) [inline CMakeLists.txt]: Do you know whether it has an effect to set `Boost_DETAILED_FAILURE_MSG` to `ON` if `REQUIRED` is used in `find_package` (see [here](https://cmake.org/cmake/help/v3.1/module/FindBoost.html))?
Anyway, it does not harm to set it!

File: test/decayAmplitude/testAmplitudeTree.cc (hunk 11 above)
Comment (bgrube) [reply test/decayAmplitude/testAmplitudeTree.cc]: Done.

File: CMakeLists.txt (hunk 12 above)
Comment (bgrube) [reply CMakeLists.txt]: Unfortunately I also don't know what this will do exactly. Thought it might be useful. Let's see...

**Comment by stefan-wallner** [comment]
Thanks a lot.

**Pull request merged**